src/mo_frontend/typing.ml

```diff
@@ -1346,6 +1346,12 @@ and infer_exp'' env exp : T.typ =
           error env typ.at "M0041" "shared function has non-async result type%a"
             display_typ_expand codom
       end
+      else if not (is_asyncE exp1) then (* local functions should be effecct free unless `async` result *)
+        begin
+          if T.(exp1.note.note_eff <> Triv) then
+            local_error env exp1.at "M0179" (*FIXME*)
```

Comment:
This is a bit gross, but I have no idea how to implement this without passing around an upper capability limit. Maybe we can find the culprit retroactively.

Reply:
Fixed by #4419!